#include "travelmoment.h"
#include "ui_travelmoment.h"
#include "clickablelabel.h"

#include <QLabel>
#include <QPushButton>
#include <QGridLayout>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFileDialog>
#include <QMessageBox>
#include <QScrollArea>
#include <QPixmap>
#include <QLayoutItem>

TravelMoment::TravelMoment(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::TravelMoment)
{
    ui->setupUi(this);
    // optional: åˆå§‹åŒ–æ˜¾ç¤ºï¼ˆè‹¥å¸Œæœ›æœ‰ç¤ºä¾‹å†…å®¹ï¼‰
    // MomentItem demo; demo.userName = "ç¤ºä¾‹"; demo.content = "æ¬¢è¿ä½¿ç”¨åŠ¨æ€æ¨¡å—"; addMoment(demo);
}

TravelMoment::~TravelMoment()
{
    delete ui;
}

// travelmoment.cpp
void TravelMoment::addMoment(const MomentItem &item)
{
    qDebug() << "æ·»åŠ åŠ¨æ€ï¼š" << item.content;

    // åˆ›å»ºå¡ç‰‡
    QWidget *card = new QWidget;
    card->setStyleSheet("background-color: #2a2a2a; border-radius: 8px; padding: 10px; margin: 5px;");

    QVBoxLayout *cardLayout = new QVBoxLayout(card);
    cardLayout->setSpacing(6);
    cardLayout->setContentsMargins(10, 10, 10, 10);

    // ç”¨æˆ·å + æ—¶é—´
    QHBoxLayout *headerLayout = new QHBoxLayout;
    QLabel *userLabel = new QLabel(item.userName);
    userLabel->setStyleSheet("color: #4fc3f7; font-weight: bold; font-size: 14px;");
    QLabel *timeLabel = new QLabel(item.publishTime.toString("yyyy-MM-dd HH:mm"));
    timeLabel->setStyleSheet("color: #999; font-size: 12px;");
    headerLayout->addWidget(userLabel);
    headerLayout->addStretch();
    headerLayout->addWidget(timeLabel);
    cardLayout->addLayout(headerLayout);

    // å†…å®¹
    QLabel *contentLabel = new QLabel(item.content);
    contentLabel->setWordWrap(true);
    contentLabel->setStyleSheet("color: white; font-size: 14px; margin-top: 5px;");
    cardLayout->addWidget(contentLabel);

    // å›¾ç‰‡ï¼ˆæ”¯æŒå¤šå¼ ï¼Œä½†å…ˆæ˜¾ç¤ºç¬¬ä¸€å¼ ï¼‰
    if (!item.images.isEmpty()) {
        ClickableLabel *imageLabel = new ClickableLabel();
        QPixmap pixmap(item.images.first());
        if (!pixmap.isNull()) {
            pixmap = pixmap.scaled(200, 200, Qt::KeepAspectRatio, Qt::SmoothTransformation);
            imageLabel->setPixmap(pixmap);
            imageLabel->setAlignment(Qt::AlignCenter);
            imageLabel->setStyleSheet("margin-top: 8px; border: 1px solid #444;");

            // ğŸ” ç‚¹å‡»å›¾ç‰‡é¢„è§ˆï¼ˆåç»­å¯æ‰©å±•ï¼‰
            imageLabel->setCursor(Qt::PointingHandCursor);
            imageLabel->setProperty("fullImagePath", item.images.first()); // å­˜å‚¨å®Œæ•´è·¯å¾„
            connect(imageLabel, &ClickableLabel::clicked, this, [this, path = item.images.first()]() {
                onImageClicked(path);
            });
        } else {
            qDebug() << "âš ï¸ å›¾ç‰‡åŠ è½½å¤±è´¥ï¼š" << item.images.first();
            QLabel *errLabel = new QLabel("å›¾ç‰‡åŠ è½½å¤±è´¥");
            errLabel->setStyleSheet("color: red; font-style: italic;");
            cardLayout->addWidget(errLabel);
        }
        cardLayout->addWidget(imageLabel);
    }

    // ç‚¹èµ/è¯„è®ºï¼ˆç®€åŒ–ï¼‰
    QHBoxLayout *actionLayout = new QHBoxLayout;
    QPushButton *likeBtn = new QPushButton(QString("ğŸ‘ %1").arg(item.likeCount));
    QPushButton *commentBtn = new QPushButton(QString("ğŸ’¬ %1").arg(item.commentCount));
    likeBtn->setFlat(true);
    commentBtn->setFlat(true);
    likeBtn->setStyleSheet("color: #ff9800; text-align: left;");
    commentBtn->setStyleSheet("color: #4caf50; text-align: left;");
    actionLayout->addWidget(likeBtn);
    actionLayout->addWidget(commentBtn);
    actionLayout->addStretch();
    cardLayout->addLayout(actionLayout);

    // ğŸ‘‡ å…³é”®ï¼šæ·»åŠ åˆ° momentListLayoutï¼ˆä¸æ˜¯ scrollAreaWidgetContents ç›´æ¥åŠ ï¼ï¼‰
    ui->momentListLayout->addWidget(card);

    // åˆ·æ–°æ»šåŠ¨åŒºï¼ˆé€šå¸¸ä¸éœ€è¦ï¼Œå› ä¸º layout è‡ªåŠ¨ç®¡ç†ï¼‰
}

void TravelMoment::refreshList()
{
    // æ¸…ç©ºæ—§çš„å¸ƒå±€é¡¹ï¼ˆå®‰å…¨åˆ é™¤ï¼‰
    QVBoxLayout* lay = new QVBoxLayout(ui->scrollAreaWidgetContents);
    ui->scrollAreaWidgetContents->setLayout(lay);
    lay->setSpacing(10);
    lay->setContentsMargins(0, 0, 0, 0);

    if (lay) {
        QLayoutItem *it;
        while ((it = lay->takeAt(0)) != nullptr) {
            if (QWidget *w = it->widget()) {
                w->deleteLater();
            } else if (QLayout *sub = it->layout()) {
                QLayoutItem *subIt;
                while ((subIt = sub->takeAt(0)) != nullptr) {
                    if (QWidget *sw = subIt->widget()) sw->deleteLater();
                    delete subIt;
                }
                delete sub;
            }
            delete it;
        }
    }

    // ç©ºçŠ¶æ€æç¤º
    if (moments.isEmpty()) {
        QLabel *empty = new QLabel("è¿˜æ²¡æœ‰åŠ¨æ€ï¼Œå¿«æ¥å‘å¸ƒç¬¬ä¸€æ¡å§ï½");
        empty->setAlignment(Qt::AlignCenter);
        empty->setStyleSheet("color:#999; padding:40px 0;");
        lay->addWidget(empty);
        lay->addStretch();
        return;
    }

    // é€æ¡åˆ›å»ºå¡ç‰‡
    for (const auto &m : moments) {
        // å¤–å±‚å¡ç‰‡
        QWidget *card = new QWidget;
        card->setObjectName("momentCard");
        // card èƒŒæ™¯ + åœ†è§’ + å†…è¾¹è·
        card->setStyleSheet(
            "#momentCard { background: #ffffff; border-radius: 10px; }"
            );
        QVBoxLayout *cardLay = new QVBoxLayout(card);
        cardLay->setContentsMargins(14,14,14,14);
        cardLay->setSpacing(10);

        // å¤´éƒ¨ï¼šå¤´åƒã€åç§°ã€æ—¶é—´
        QHBoxLayout *header = new QHBoxLayout;
        header->setSpacing(10);

        // å¤´åƒï¼ˆåœ†å½¢ç°è‰²å ä½ï¼‰
        QLabel *avatar = new QLabel;
        avatar->setFixedSize(44,44);
        avatar->setStyleSheet("background:#e6e6e6; border-radius:22px; color:#888; font-size:12px;");
        avatar->setAlignment(Qt::AlignCenter);
        avatar->setText("å¤´åƒ");

        // åç§° + æ—¶é—´å‚ç›´æ’åˆ—
        QVBoxLayout *nameTime = new QVBoxLayout;
        nameTime->setSpacing(2);
        QLabel *name = new QLabel(m.userName);
        name->setStyleSheet("color:#222; font-weight:600; font-size:14px;");
        QLabel *time = new QLabel(m.publishTime.toString("MM-dd hh:mm"));
        time->setStyleSheet("color:#999; font-size:12px;");
        nameTime->addWidget(name);
        nameTime->addWidget(time);

        header->addWidget(avatar);
        header->addLayout(nameTime);
        header->addStretch();

        cardLay->addLayout(header);

        // æ­£æ–‡ï¼ˆç¡®ä¿æ·±è‰²æ–‡å­—å¯è§ï¼‰
        QLabel *content = new QLabel;
        content->setWordWrap(true);
        content->setText(m.content.isEmpty() ? QString() : m.content);
        content->setStyleSheet("color:#222; font-size:14px; line-height:1.4;");
        // è®©å†…å®¹ä¸å¡ç‰‡ä¿æŒä¸€å®šæœ€å°é«˜åº¦ï¼ˆç©ºæ–‡æœ¬æ—¶ä¸è¦å®Œå…¨å¡Œé™·ï¼‰
        content->setMinimumHeight(20);
        cardLay->addWidget(content);

        // å›¾ç‰‡ç½‘æ ¼ï¼ˆå¦‚æœæœ‰ï¼‰
        if (!m.images.isEmpty()) {
            QGridLayout *grid = new QGridLayout;
            grid->setSpacing(6);
            int cnt = qMin(9, m.images.size());
            for (int i = 0; i < cnt; ++i) {
                QLabel *img = new QLabel;
                img->setFixedSize(100,100);
                img->setScaledContents(true);
                QPixmap p(m.images[i]);
                if (!p.isNull()) {
                    img->setPixmap(p.scaled(img->size(), Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation));
                } else {
                    img->setStyleSheet("background:#f0f0f0; border:1px solid #eee;");
                    img->setText("å›¾ç‰‡");
                    img->setAlignment(Qt::AlignCenter);
                }
                grid->addWidget(img, i/3, i%3);
            }
            cardLay->addLayout(grid);
        }

        // åº•éƒ¨æ“ä½œæ ï¼ˆç‚¹èµ/è¯„è®ºï¼‰
        QFrame *sep = new QFrame;
        sep->setFrameShape(QFrame::HLine);
        sep->setFrameShadow(QFrame::Sunken);
        sep->setStyleSheet("color:#eee;");
        cardLay->addWidget(sep);

        QHBoxLayout *footer = new QHBoxLayout;
        footer->setSpacing(16);

        // like æŒ‰é’®é£æ ¼
        QString likeText = (m.liked ? QString("â™¥ ") : QString("â™¡ ")) + QString::number(m.likeCount);
        QPushButton *likeBtn = new QPushButton(likeText);
        likeBtn->setFlat(true);
        likeBtn->setCursor(Qt::PointingHandCursor);
        likeBtn->setStyleSheet("QPushButton{color:#ff4d4f; font-size:13px;} QPushButton:pressed{opacity:0.8;}");

        QPushButton *commentBtn = new QPushButton(QString("è¯„è®º ") + QString::number(m.commentCount));
        commentBtn->setFlat(true);
        commentBtn->setCursor(Qt::PointingHandCursor);
        commentBtn->setStyleSheet("QPushButton{color:#333; font-size:13px;}");

        footer->addStretch();
        footer->addWidget(likeBtn);
        footer->addWidget(commentBtn);

        cardLay->addLayout(footer);

        // è¿æ¥ lambdaï¼ˆä¿æŒå’Œæ•°æ®åŒæ­¥ï¼‰
        connect(likeBtn, &QPushButton::clicked, this, [this, id = m.id, likeBtn]() {
            for (auto &mm : moments) {
                if (mm.id == id) {
                    mm.liked = !mm.liked;
                    mm.likeCount += mm.liked ? 1 : -1;
                    QString txt = (mm.liked ? QString("â™¥ ") : QString("â™¡ ")) + QString::number(mm.likeCount);
                    likeBtn->setText(txt);
                    break;
                }
            }
        });

        connect(commentBtn, &QPushButton::clicked, this, [this, id = m.id]() {
            QMessageBox::information(this, "è¯„è®º", "ç‚¹å‡»äº†è¯„è®ºï¼ŒåŠ¨æ€IDï¼š" + QString::number(id));
        });

        // æ·»åŠ å¡ç‰‡åˆ°åˆ—è¡¨
        lay->addWidget(card);
    }

    // åº•éƒ¨æ‹‰ä¼¸
    if (lay) lay->addStretch();
}


void TravelMoment::on_publishBtn_clicked()
{
    QString content = ui->contentEdit->toPlainText().trimmed();
    if (content.isEmpty()) {
        QMessageBox::warning(this, "æç¤º", "è¯·è¾“å…¥å†…å®¹ï¼");
        return;
    }

    MomentItem newItem;
    newItem.id = nextId++;
    newItem.content = content;
    newItem.images = selectedImages;
    newItem.publishTime = QDateTime::currentDateTime();
    newItem.userName = "å½“å‰ç”¨æˆ·"; // TODO: æ›¿æ¢ä¸ºçœŸå®ç”¨æˆ·å

    moments.append(newItem);
    addMoment(newItem);

    // æ¸…ç©º
    ui->contentEdit->clear();
    selectedImages.clear();
    ui->imagePreview->setText("å·²é€‰æ‹© 0 å¼ å›¾ç‰‡");
}

void TravelMoment::on_selectImageBtn_clicked()
{
    QStringList paths = QFileDialog::getOpenFileNames(
        this,
        "é€‰æ‹©å›¾ç‰‡",
        "",
        "å›¾ç‰‡æ–‡ä»¶ (*.png *.jpg *.jpeg *.bmp)"
        );

    if (paths.isEmpty()) return;

    // é™åˆ¶æœ€å¤š9å¼ 
    for (const QString &path : paths) {
        if (selectedImages.size() >= 9) break;
        if (!selectedImages.contains(path)) {
            selectedImages.append(path);
        }
    }

    ui->imagePreview->setText(QString("å·²é€‰æ‹© %1 å¼ å›¾ç‰‡").arg(selectedImages.size()));
}

void TravelMoment::onLikeClicked(int id)
{
    // åˆ‡æ¢å¯¹åº” id çš„ç‚¹èµçŠ¶æ€å¹¶åˆ·æ–°åˆ—è¡¨æ˜¾ç¤º
    for (auto &mm : moments) {
        if (mm.id == id) {
            mm.liked = !mm.liked;
            mm.likeCount += mm.liked ? 1 : -1;
            break;
        }
    }
    refreshList();
}

void TravelMoment::onCommentClicked(int id)
{
    // ç›®å‰ç¤ºä¾‹åªå¼¹å‡ºæç¤ºæ¡†ï¼›ä»¥åå¯å¼¹å‡ºè¯„è®ºçª—å£å¹¶æŠŠè¯„è®ºå†™å…¥ DB
    QMessageBox::information(this, tr("è¯„è®º"), tr("ç‚¹å‡»äº†è¯„è®ºï¼ŒåŠ¨æ€IDï¼š%1").arg(id));
}

void TravelMoment::onImageClicked(const QString &path)
{
    showImagePreview(path);
}

void TravelMoment::showImagePreview(const QString &imagePath)
{
    QPixmap pixmap(imagePath);
    if (pixmap.isNull()) return;

    // ğŸ‘‰ ä½¿ç”¨ ClickableLabel è€Œä¸æ˜¯ QLabel
    ClickableLabel *preview = new ClickableLabel();
    preview->setPixmap(pixmap.scaled(600, 600, Qt::KeepAspectRatio, Qt::SmoothTransformation));
    preview->setWindowTitle("å›¾ç‰‡é¢„è§ˆ - ç‚¹å‡»å…³é—­");
    preview->setWindowFlags(Qt::Window | Qt::FramelessWindowHint);
    preview->setStyleSheet("background: black;");
    preview->setAlignment(Qt::AlignCenter);
    preview->setAttribute(Qt::WA_DeleteOnClose);

    // âœ… è¿æ¥ç‚¹å‡»ä¿¡å·
    connect(preview, &ClickableLabel::clicked, preview, &QWidget::close);

    preview->show();
}
